### PR TITLE
Use Ignite's interrupt api in MonaiAlgo

### LINKED ISSUE #4554: Add interrupt/terminate option to Trainers & Evaluators
**Is your feature request related to a problem? Please describe.**
It should be possible to abort/finalize a running Trainer by calling the API (rather than ctr+C). This will be helpful if the Trainer needs to be executed remotely, such as in federated learning (FL) scenarios.

**Describe the solution you'd like**
Add abort() and finalize() functions to the [Trainer](https://github.com/Project-MONAI/MONAI/blob/d34fa14eed3191126fc1bc1a9751e69e0396f78b/monai/engines/trainer.py#L45) class (or potentially its base class). Note, finalize() should terminate the training completely, while abort() should allow later continue of where it was aborted(), by calling run() again.

For example, an ignite-based Trainer support abort() and finalize() calls could be implemented as such (Currently used in MONAI-FL's [MonaiAlgo class](https://github.com/Project-MONAI/monai-fl/blob/main/monaifl/monai_algo.py); private repo - contact me if you need access)
```
    def abort(self):
        self.trainer.terminate()
        # save current iteration for next round
        setattr(self.trainer.state, "dataloader_iter", self.trainer._dataloader_iter)

        if self.trainer.state.iteration % self.trainer.state.epoch_length == 0:
            # if current iteration is end of 1 epoch, manually trigger epoch completed event
            self.trainer._fire_event(Events.EPOCH_COMPLETED)

    def finalize(self):
        self.trainer.terminate()
```

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by holgerroth**
Fixes #4554. 

Requires latest Ignite RC.

Also improves ClientAlgo docstrings formatting.

### Description
A few sentences describing the changes proposed in this pull request.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/fl/client/client_algo.py [hunk 1]
```diff
@@ -20,43 +20,69 @@ class ClientAlgo(abc.ABC):
...
+        Train network and produce new network from train data.
+
+        Args:
+            data: ExchangeObject containing current network weights to base training on.
+            extra: optional extra information
+
+        Returns:
+            None
         """
         raise NotImplementedError
```
Comment (Nic-Ma) [inline monai/fl/client/client_algo.py:71]: I would suggest to add some error message, like: https://github.com/Project-MONAI/MONAI/blob/dev/monai/engines/workflow.py#L293

Thanks.

File: monai/fl/client/client_algo.py [hunk 2]
```diff
@@ -20,43 +20,69 @@ class ClientAlgo(abc.ABC):
...
+    of the class object.
     """
 
     def initialize(self, extra=None):
-        """call to initialize the ClientAlgo class"""
+        """
+        Call to initialize the ClientAlgo class
+
+        Args:
+            extra: optional extra information, e.g. dict of `ExtraItems.CLIENT_NAME` and/or `ExtraItems.APP_ROOT`
```
Comment (Nic-Ma) [inline monai/fl/client/client_algo.py:36]: Should we require the type of `extra` in the base class?
For example: `Optional[dict]`.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @holgerroth @wyli ,

I think we should update the `min_version` of ignite in the monai_algo.py file to `0.4.10`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/fl/client/monai_algo.py#L41
Other places keep the original min version.
Then when ignite released 0.4.10 next Monday, let's update the ignite version in all the docker and CI configs.
Now the min required version is `0.4.4`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/config/deviceconfig.py#L261
And the docker and CI version is `0.4.9`:
https://github.com/Project-MONAI/MONAI/blob/dev/requirements-dev.txt#L3

What do you think?

Thanks.

**Comment by holgerroth** [comment]
> Hi @holgerroth @wyli ,
> 
> I think we should update the `min_version` of ignite in the monai_algo.py file to `0.4.10`: https://github.com/Project-MONAI/MONAI/blob/dev/monai/fl/client/monai_algo.py#L41 Other places keep the original min version. Then when ignite released 0.4.10 next Monday, let's update the ignite version in all the docker and CI configs. Now the min required version is `0.4.4` for all the places: https://github.com/Project-MONAI/MONAI/blob/dev/monai/config/deviceconfig.py#L261 And the docker and CI version is `0.4.9`: https://github.com/Project-MONAI/MONAI/blob/dev/requirements-dev.txt#L3
> 
> What do you think?
> 
> Thanks.

Actually, we don't need to import Ignite anymore in this class. I removed it.

File: monai/fl/client/client_algo.py (hunk 1 above)
Comment (holgerroth) [reply monai/fl/client/client_algo.py:71]: Good point. I added it.

File: monai/fl/client/client_algo.py (hunk 2 above)
Comment (holgerroth) [reply monai/fl/client/client_algo.py:37]: Thanks. I added the typing info.

**Comment by vfdev-5** [comment]
FYI, we just released [0.4.10](https://github.com/pytorch/ignite/releases/tag/v0.4.10), so you can check and also update MONAI deps.

**Review by Nic-Ma — APPROVED** [review]
I have updated the ignite dependency in the PR.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Could you please help double confirm it?

Thanks in advance.

**Comment by wyli** [comment]
> Hi @wyli ,
> 
> Could you please help double confirm it?
> 
> Thanks in advance.

I see there are docs-test errors: https://github.com/Project-MONAI/MONAI/runs/8197608122?check_suite_focus=true

File: monai/fl/client/client_algo.py [hunk 3]
```diff
@@ -20,60 +21,88 @@ class ClientAlgo(abc.ABC):
...
+
+        Args:
+            data: ExchangeObject containing current network weights to base training on.
+            extra: optional extra information
+
+        Returns:
+            None
         """
-        raise NotImplementedError
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
```
Comment (vfdev-5) [inline monai/fl/client/client_algo.py:72]: I wonder whether this error will be ever seen as method is decorated with `@abc.abstractmethod` ?
As we can't instantiate `ClientAlgo`, the only way as far as I can see to raise this error is to call the method directly as static : `ClientAlgo.train(None, None)`...

Comment (Nic-Ma) [reply monai/fl/client/client_algo.py:71]: Hi @vfdev-5 ,

That's a good point.
I removed the decorator in the latest commit, it's same as the `Workflow` class now:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/engines/workflow.py#L280
And tested as below:
```py
>>> from monai.fl.client.client_algo import ClientAlgo
>>>
>>>
>>> class Test(ClientAlgo):
...     pass
...
>>> t = Test()
>>> t.train(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/workspace/data/medical/MONAI/monai/fl/client/client_algo.py", line 71, in train
    raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
NotImplementedError: Subclass Test must implement this method.
```

Thanks.

Comment (vfdev-5) [reply monai/fl/client/client_algo.py:71]: I see. So, in this case is there a reason to derive `ClientAlgo` from `abc.ABC` ?

Comment (Nic-Ma) [reply monai/fl/client/client_algo.py:70]: You are right, also removed `ABC`.

Thanks.

**Pull request merged**